### PR TITLE
items: add new fields to inventory list

### LINKED ISSUE #2704: Add some fields to the Inventory list
**How it works**

In the inventory list, some fields are missing to help librarians in the collection management.

**Improvement suggestion**

Please add the following fields :  
From the document records 
- Identifiers : ISBN and ISSN
- Series statement : _text (so we can see all the parts and numbering)
- Provision Activity : Agent (in the Statement)
- Provision Activity : startDate and endDate
- Edition statement : _text
- Masked : Yes/No
- Local fields of the organisation

From the item records and loans
- Creation date of the item
- Temporary circulation category : expiry date
- All the notes type (notes.type:"acquisition_note" ; "binding_note" ; "condition_note" ; "patrimonial_note" ; "provenance_note"). [in addition to the one that are currently in the inventory]
- Masked : Yes/No
- PAC code
- Number of renewals
- Local fields

Check the field called "created" in the current CSV export
- if it is the creation of the item, put the field in the beginning of the columns
- if it is another creation date, rename the field

See current bug at exporting loan count: https://github.com/rero/rero-ils/issues/2793

Nice to have : Be able to choose the fields we need when exporting

### LINKED ISSUE #2793: Inventory list export: some columns counts only use active loans
### How it works

When exporting a CSV from an inventory list (pro UI), some columns only count the values from currently active loans (`ITEM_AT_DESK`, `ITEM_ON_LOAN`, `ITEM_IN_TRANSIT_FOR_PICKUP`, `ITEM_IN_TRANSIT_TO_HOUSE`) for an item where they should be more global:

* `loans_count` column is always binary (either 0 or 1). It counts the number of active loans. This is not needed for users; they can get this information from the `status` column.
* `last_transaction_date` only displays something if there is an active loan for the item. Furthermore it displays the last date of the document `operation log`. This should be from the item/loan operation log. Librarians want to see when the item was last circulated, whether the loan is still active or not. If an item has been `ON_SHELF` for six months, they want to know.

### Improvement suggestion

* `loans_count` : in this column we want to display the number of loans for an item, i.e. the number of checkout transactions this item has had in its entire lifetime (like for the stats module). 
* `last_transaction_date` : for the values to be coherent with this label, this should show the latest transaction date whether the item still has an active loan or not. 

### Nice to have

Add one more column in the export (after `loans_count`) that also displays the number of renewal transaction that the item has had in its lifetime. Call this (`renewals_count`).

### PR DESCRIPTION
**Pull request by BadrAly**
* Fixes #2704
* Fixes #2793

Co-Authored-by: Aly Badr <aly.badr@rero.ch>

# How to Test:

- run the `Inventory list` report  for a given instance.
- the fields extracted from the `document` record are now prefixed with the string `document_`
- the fields extracted from the `item` record are now prefixed with the string `item_`
- the `local_fields` are now extracted for both `documents` and `items`
- all the fields mentioned in the issue https://github.com/rero/rero-ils/issues/2704 are added
- loan information are correctly extracted according to https://github.com/rero/rero-ils/issues/2793

### CODE REVIEW

File: rero_ils/modules/items/serializers/collector.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+            org_pid = csv_data['item_org_pid']
+            if local_fields := LocalField.get_local_fields_by_resource(
+                    resource_type, resource_pid, organisation_pid=org_pid):
+                for field, num in itertools.product(
+                        local_fields, range(1, 11)):
+                    field_name = f'{resource_type}_local_field_{num}'
+                    if field_data := field.get(
+                            'fields', {}).get(f'field_{num}'):
+                        csv_data[field_name] = \
+                            ' '.join(filter(None, field_data))
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector.py:296]: `filter(None, ...)` seems not necessary here

File: rero_ils/modules/items/serializers/collector.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+        ]
+        loans_agg = A('filter', terms={'loan.trigger': triggers},
+                      aggs=dict(
+            item_pid=A('terms',
+                       field='loan.item.pid',
+                       aggs=dict(last_op=A('max', field='date')))))
+        query.aggs.bucket('loans', loans_agg)
+        # query execution
+        result = query.execute()
+        # dump outout into a dict
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector.py:345]: typo: outout => output

File: rero_ils/modules/items/serializers/collector.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        :param item_pids: item pids.
+        :return dict of item_pids having the checkouts, renewals,
+                last_transaction_dates, last_checkouts data.
+        """
+        def add_pid(loans, pid):
+            """Add pid to loans."""
+            if pid not in loans:
+                loans[pid] = {}
+
+        # initial es query
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector.py:314]: maybe a little bit explanation about this query could be nice. (with a sample output exemple ?)

File: rero_ils/modules/items/serializers/csv.py [hunk 4]
```diff
@@ -18,49 +18,23 @@
...
-    'art',
-    'ive',
-    'chr',
-    'aut',
-    'arc',
-    'fmk',
-    'pra',
-    'csl'
-]
+from ..serializers.collector import Collecter
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/csv.py:34]: `from .collector` seems suffisant ;-)

File: rero_ils/modules/items/serializers/csv.py [hunk 5]
```diff
@@ -215,80 +80,23 @@ def get_loans_by_item_pids(item_pids):
...
-                    except Exception as err:
-                        current_app.logger.error(
-                            'ERROR in csv serializer: '
-                            '{message} on document: {pid}'.format(
-                                message=err,
-                                pid=hit['document']['pid'])
-                        )
-
+                    csv_data = Collecter.get_item_data(hit)
+                    _process_item_types_libs_locs(self, csv_data)
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/csv.py:91]: `self._process_item_types_libs_locs()`

File: tests/api/test_serializers.py [hunk 6]
```diff
@@ -176,15 +176,37 @@ def test_items_serializers(
...
+        'item_local_field_2', 'item_local_field_3', 'item_local_field_4',
+        'item_local_field_5', 'item_local_field_6', 'item_local_field_7',
+        'item_local_field_8', 'item_local_field_9', 'item_local_field_10',
+        'issue_status', 'issue_status_date', 'issue_claims_count',
+        'issue_expected_date', 'issue_regular', 'item_checkouts_count',
+        'item_renewals_count', 'last_transaction_date', 'checkout_date'
+    ]
+    for field in fields:
+        assert field in data
+    assert 'item_pid' in data
```
Comment (zannkukai) [inline tests/api/test_serializers.py:209]: why not adding "item_pid" into the field list ?

File: rero_ils/modules/items/serializers/collector.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+        'ive',
+        'chr',
+        'aut',
+        'arc',
+        'fmk',
+        'pra',
+        'csl'
+    ]
+
+    @classmethod
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/items/serializers/collector.py:64]: done when possible

**Comment by PascalRepond** [comment]
## Tests

* [ ] document_series_statement and `document_edition_statement` are not exported
* [ ] `document_local_fields` not exported 
* [ ] `temporary_item_type` not exported

**Comment by benerken** [comment]
Great job ! Thank you.

- [ ] please move the first 2 columns (item_pid and item_create_date) and try group them with the other Item columns.
      => The purpose is to get Document info, then Item info, then loan infos.

**Review by benerken — APPROVED** [review]

File: rero_ils/modules/items/serializers/csv.py [hunk 8]
```diff
@@ -72,141 +46,34 @@ def serialize_search(self, pid_fetcher, search_result, links=None,
...
+                if temporary_item_type_pid := csv_data.pop(
+                        'temporary_item_type_pid', None):
+                    csv_data['temporary_item_type'] = self.get_resource(
+                        ItemType, temporary_item_type_pid).get('name')
+                # library
+                csv_data['item_library_name'] = self.get_resource(
+                    Library, csv_data.pop('item_library_pid')).get('name')
+                # location
+                csv_data['item_location_name'] = self.get_resource(
+                    Location, csv_data.pop('item_location_pid')).get('name')
```
Comment (lauren-d) [inline rero_ils/modules/items/serializers/csv.py:68]: At this time, using resource cache isn't a good idea.
I think it's better to load all resources of a type in one call.
This method must be refactorized when it is possible to load all resources in the cached class.
At least, could you add a "TODO" in code?

**Review by lauren-d — APPROVED** [review]

**Pull request merged**